### PR TITLE
Support character move forward and backward with different speed.

### PR DESCRIPTION
**Pull request by watsonsong**
Support character move forward and backward with different speed.
It let the backword movement can be slower.

### CODE REVIEW

File: Source/ALS/Public/State/AlsGroundedState.h [hunk 1]
```diff
@@ -22,6 +22,9 @@ struct ALS_API FAlsVelocityBlendState
 	UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = "ALS")
 	uint8 bReinitializationRequired : 1 {true};
 
+	UPROPERTY(EditAnywhere, BlueprintReadWrite, Category = "ALS", Meta = (ClampMin = 0, ClampMax = 1))
+	FVector2D RelativeDirection{FVector2D::ZeroVector};
```
Comment (Sixze) [inline Source/ALS/Public/State/AlsGroundedState.h:26]: Seems unused.

File: Source/ALS/Public/Settings/AlsMovementSettings.h [hunk 2]
```diff
@@ -64,16 +70,18 @@ class ALS_API UAlsMovementSettings : public UDataAsset
...
 	{
-		return WalkSpeed;
+		return FMath::GetMappedRangeValueClamped(FVector2d(-1, 0),
+			FVector2d(WalkBackwardSpeed, WalkForwardSpeed), LocalInputVector.X);
 	}
 
 	if (Gait == AlsGaitTags::Running)
 	{
-		return RunSpeed;
+		return FMath::GetMappedRangeValueClamped(FVector2d(-1, 0),
```
Comment (Sixze) [inline Source/ALS/Public/Settings/AlsMovementSettings.h:83]: Use floating point literals, such as `0.0f`.

File: Source/ALS/Private/AlsCharacterMovementComponent.cpp [hunk 3]
```diff
@@ -947,23 +968,26 @@ float UAlsCharacterMovementComponent::CalculateGaitAmount() const
...
 	}
 
-	if (Speed <= GaitSettings.RunSpeed)
+	if (Speed <= GaitSettings.RunForwardSpeed)
 	{
 		static const FVector2f GaitAmount{1.0f, 2.0f};
 
-		return FMath::GetMappedRangeValueClamped({GaitSettings.WalkSpeed, GaitSettings.RunSpeed}, GaitAmount, Speed);
+		const float WalkSpeed = GetWalkSpeed();
+		const float RunSpeed = GetRunSpeed();
```
Comment (Sixze) [inline Source/ALS/Private/AlsCharacterMovementComponent.cpp:984]: Use uniform initialization.

File: Source/ALS/Private/AlsCharacter.cpp [hunk 4]
```diff
@@ -991,12 +996,12 @@ FGameplayTag AAlsCharacter::CalculateActualGait(const FGameplayTag& MaxAllowedGa
 	// different from the desired gait or max allowed gait. For instance, if the max allowed gait becomes
 	// walking, the new gait will still be running until the character decelerates to the walking speed.
 
-	if (LocomotionState.Speed < AlsCharacterMovement->GetGaitSettings().WalkSpeed + 10.0f)
+	if (LocomotionState.Speed < AlsCharacterMovement->GetWalkSpeed() + 10.0f)
```
Comment (Sixze) [inline Source/ALS/Private/AlsCharacter.cpp:999]: I think this is wrong. You should use `WalkForwardSpeed` instead.

**Comment by Sixze** [comment]
Have you tested this in multiplayer? I have doubts that it will work correctly there.

File: Source/ALS/Private/AlsCharacter.cpp (hunk 4 above)
Comment (watsonsong) [reply Source/ALS/Private/AlsCharacter.cpp:999]: Is means calculate the gait alway use the forwad speed, even if the actor is walking backward?

**Comment by watsonsong** [comment]
> Have you tested this in multiplayer? I have doubts that it will work correctly there.

I test in PIE mode with 'Play as Listen Server' and 'Play as Client'(pure dedicated server). It seems animation and speed is correct when backward.

File: Source/ALS/Private/AlsCharacter.cpp [hunk 5]
```diff
@@ -963,6 +963,11 @@ void AAlsCharacter::RefreshGait()
 
 	AlsCharacterMovement->SetMaxAllowedGait(MaxAllowedGait);
 
+	const auto RelativeVelocityDirection{
+		LocomotionState.RotationQuaternion.UnrotateVector(LocomotionState.Velocity).GetSafeNormal()
+	};
+	AlsCharacterMovement->SetRelativeVelocityDirection(RelativeVelocityDirection);
```
[the remark this answers is not in the document]
Comment (watsonsong) [reply Source/ALS/Private/AlsCharacter.cpp:969]: Is the 'CalcVelocity' a good place to calculate this vector and update new max walk speed?

**Comment by Sixze** [comment]
> I test in PIE mode with 'Play as Listen Server' and 'Play as Client'(pure dedicated server). It seems animation and speed is correct when backward.

But what about network simulation?

File: Source/ALS/Private/AlsCharacter.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sixze) [reply Source/ALS/Private/AlsCharacter.cpp:969]: You may not even need to cache this value, just create a function in the character movement component that will calculate this value on the fly and use it.

**Comment by onx2** [comment]
I'm handling this in the `Input_OnMove` function, as well as strafe speed. In my use case, I don't mind the sprint being a jog while moving backward, rather than the character turning into that direction at full speed.

```c++
void AAlsCharacterExample::Input_OnMove(const FInputActionValue& ActionValue)
{
	const auto Value{UAlsMath::ClampMagnitude012D(ActionValue.Get<FVector2D>())};

	const auto ForwardDirection{UAlsMath::AngleToDirectionXY(UE_REAL_TO_FLOAT(GetViewState().Rotation.Yaw))};
	const auto RightDirection{UAlsMath::PerpendicularCounterClockwiseXY(ForwardDirection)};

	float XSpeed = Value.X;
	float YSpeed = Value.Y;
	const bool bBackward = YSpeed < 0;
	const bool bStrafe = YSpeed == 0 && XSpeed != 0;
	
	if (bBackward)
	{
		XSpeed *= 0.5f;
		YSpeed *= 0.5f;
	}
	else if (bStrafe)
	{
		XSpeed *= 0.75f;
	}
	
	AddMovementInput(ForwardDirection * YSpeed + RightDirection * XSpeed);
}
```

**Comment by watsonsong** [comment]
> I'm handling this in the `Input_OnMove` function, as well as strafe speed. In my use case, I don't mind the sprint being a jog while moving backward, rather than the character turning into that direction at full speed.
> 
> ```c++
> void AAlsCharacterExample::Input_OnMove(const FInputActionValue& ActionValue)
> {
> 	const auto Value{UAlsMath::ClampMagnitude012D(ActionValue.Get<FVector2D>())};
> 
> 	const auto ForwardDirection{UAlsMath::AngleToDirectionXY(UE_REAL_TO_FLOAT(GetViewState().Rotation.Yaw))};
> 	const auto RightDirection{UAlsMath::PerpendicularCounterClockwiseXY(ForwardDirection)};
> 
> 	float XSpeed = Value.X;
> 	float YSpeed = Value.Y;
> 	const bool bBackward = YSpeed < UE_SMALL_NUMBER;
> 	
> 	const bool bStrafe = FMath::IsNearlyZero(YSpeed) && !FMath::IsNearlyZero(XSpeed);
> 	
> 	if (bBackward)
> 	{
> 		XSpeed *= 0.5f;
> 		YSpeed *= 0.5f;
> 	}
> 	else if (bStrafe)
> 	{
> 		XSpeed *= 0.75f;
> 	}
> 	
> 	AddMovementInput(ForwardDirection * YSpeed + RightDirection * XSpeed);
> }
> ```

This is a good method. Simple enough and not Intrusion. But is it safe for dedicated server. Can the client may modify the backward and strafe speed, and the server can not detect it?

**Comment by onx2** [comment]
@watsonsong I'm fairly new to game development, especially networked game development, so I don't think I can speak to that with 100% certainty. However, my intuition tells me that it wouldn't be a problem because if they were able to modify this code than no matter what logic we put here there could be issues. For example, if they somehow got access to modify this code freely without my additions, what is stopping them from changing it to:

```c++
AddMovementInput(ForwardDirection * 99999999 + RightDirection * 9999999);
```

**Comment by watsonsong** [comment]
I means the only safe place is our dedicated server. And the client may be modified by any way. In the bad case, the cheater can make move backward as the speed as the forward movement.

**Comment by onx2** [comment]
I'm not sure I follow, as this is client-side code, so there is the same potential for cheating with or without a multiplier. You should always run validation on the server.

**Comment by watsonsong** [comment]
Yes that what I mean. The server does not known anything about the speed multiplier. So it possible to cheating with it.
So I think It may better the limit the speed on the movement logic.

**Pull request merged**